### DISCUSSION TITLE
Native support for bidirectional 1-wire-protocol (EX bus?)

### DISCUSSION THREAD

**Discussion by offer-shmuely** [ExpressLRS/ExpressLRS#1979] [Ideas]
I like express lrs very much, it's reliability, penetration, the ability to run on many radio (in oppose to access protocol that i use)
However it is currently optimized to model with flight controller, plane or quad with flight controller.

One on the issue i am straggler with, is the 4 wires connection, the rx/tx
When i put on a quad, there is no issue, it close to the FC, and I have to solder anyway.
But when i try to use it on plane and heli, i like the receiver to be interchangeable, so i solder pin header to it, but it very hard to find a servo extender wires with 4 wires, if i like to connect it to a heli FC, and put it 30cm away, i found myself wondering how? It's complicated, should i build a 30cm servo extender with 4 wires? 
When there will be bus extender for elrs, like the SB-20 of frsky, or the powerbox,how do we connect two receivers? By soldering? On 4000$ dollar plane?
How do we take on of the two receivers to the tail, 150cm away?

I feel that it will be more convenions on a day to day use to have a wire protocol of 3 wires installed of 4, so a regular servo extender can be use.

Assumption: 
on plane we need only low refresh rate, 50Hz-100Hz is good enough 

On plane a latency of 10-15mili sec is ok as well, even for 3d plane, we does not need 3mili and below.

On plane we prefer no soldering

On plane we will expect more than one receiver for redundancy (hardware failures, receiver hide behind karbon)

On plane the receiver will not always be close to it's pwm expander box, and will need extenders

I like to know, if it possible to create a wire protocol like the f.port(frsky), or sbus2(futaba), that is half duplex wire protocol to let the elrs take the market share it deserve.... the first place.

Am i the only one that feel like this, is there other that feel the same?

**Comment by offer-shmuely** [comment]
I would like to give explanation to the vision I presented above:

some heli flight controller allow two receiver, they use only 3 wires, and they can implement a new protocol if there will be a demand
![image](https://user-images.githubusercontent.com/7026911/205975076-677ffd86-7e22-48ce-9cfe-649878a96605.png)

--- 
In expensive plane we like (in the future) to do something like this, but with elrs (to throw the ACCST & ACCESS totally)
again, harder to do with 4 wires

![image](https://user-images.githubusercontent.com/7026911/205977913-f608c472-2c47-46a0-baff-11c60f8fca28.png)

---
Jhon from rcvideoreview did something like this, with current available hardware, but he had to use an SBUS, and he did not had a telemetry, note the easy usage of regular servo wires, almost no soldering, so if he liked to take one receiver to another area on the plane, it is simple since servo extender are everywhere, how do we do it with 4 wire? it becoming a DIY project, not a wide spread solution that we all what.
![image](https://user-images.githubusercontent.com/7026911/205976152-2d6a39f8-198f-4e7b-b346-5737e2ca2494.png)

**Comment by rueckwaertsflieger** [comment]
Basically, changing a bidirectional 2-wire link to a 1-wire link or 1-wire bus implies using another protocol. You are asking for a ELRS over 1-wire implementation. Well, it would be a CRSF 1-wire implementation, since ELRS is the protocol used over-the-air and CRSF is the protocol for intra-model and -radio communication, used in the ELRS ecosystem.
Read about protocols. Oscar Liang has made a short overview. Find the difference between protocols "over-air" and "baseband", Oscar is naming Tx and Rx protocol respectively (Which is quite inaccurate). https://oscarliang.com/rc-protocols/
ELRS in first place is an over-the air-protocol. But the ecosystem around ELRS allows for different protocols - theoretically. So, from the time on, developers and thus ELRS receivers would provide a baseband 1-wire-protocol in addition to CRSF by configuration, you would get your solution.
To my knowledge today the following baseband bidirectional 1-wire protocols for RC use defined are:
SBUS2
FPort (patented)
SRXL/SRXL2
All other baseband bidirectional protocols need 2 wires at least as is the case for e.g. Canbus, I2C....

SBUS is coming. However, this will be SBUS, not SBUS2.
Take a look at: https://github.com/ExpressLRS/ExpressLRS/pull/2094

**Reply by skelia-systems** [reply]
Hey, i need to connect the flight controller directly to the RadioMaster TX12 via JR-Bay pins for some reason
Using either SBUS or CRSF protocol
I just need to connect it to the FC directly and to be able to see the channels are responding on the Receiver tab in Betaflight
I use FC SpeedyBee F405V4

What hardware do i need beetween and what to change in Betaflight CLI programatically to make it work?
Thank you

**Comment by brandonrc** [comment]
Well we will soon support sbus officially. If you build with master you can use SBUS mode.

**Comment by offer-shmuely** [comment]
I do accept you clarification, thanks.
The over the air protocol is great, i do not want to change it.
The outside protocol between the receiver to the peripheral is what holding adoption rate (in my opinion) and keep only transmitter with expreeLrs.

that side i liked to be enhanced, especially for planes, but for heli as well.
For me, i need a good telemetry (crsf is too limited currently) 
And good eco system of peripherals, the best one are Jeti, so loved to see Jeti EX-bus implemented (it is also 3 wires)

From their standard: 

https://www.jetimodel.cz/uloz/c6d00c58-78ab-4c4d-9f15-34e84523479e.pdf

 EX Bus is a standard of serial data transmission, designed primarily for transmission of operational information between a receiver and other devices connected to it. EX Bus replaces standard of channel values transmission in the form of PPM, and also provides security features, faster response, and bidirectional communication (transmission of EX telemetry). EX Bus also enables transmission of information for remote configuration of connected devices via the DC/DS transmitters.

**Comment by rueckwaertsflieger** [comment]
OK, did not know about EX Bus.
So this discussion is tending towards another 1-wire protocol, one more besides SBUS. However, I suggest to either alter the headline of this discussion or even better, open a new one. Kind of "Native support for Ex Bus" or "Native support for bidirectional 1-wire-protocol".

**Comment by pkendall64** [comment]
As @brandonrc mentioned, we have SBUS coming. As part of that work, the entire serial protocol stack was refactored to allow other protocols to be easily added, so if there are open documents on EX Bus and they are not encumbered by a patent in any way it could be added fairly easily. There may be some extra changes required as it's a bidirectional single wire protocol, but that can be worked through.

**Reply by offer-shmuely** [reply]
Hi
Yes, there is a documentation for the Jeti ex-bus
Here it is:
https://www.jetimodel.cz/uloz/c6d00c58-78ab-4c4d-9f15-34e84523479e.pdf

**Comment by offer-shmuely** [comment]
Still hope it will happen one day…
